[Synthetics] Private locations (#5162)
DataDog/documentation: content/en/synthetics/private_locations.md

kind: documentation
description: Identify incoming Synthetics requests
further_reading:
- link: "synthetics/"
  tag: "Documentation"
  text: "Manage your checks"
- link: "synthetics/browser_tests"
  tag: "Documentation"
  text: "Configure a Browser Test"
- link: "synthetics/api_tests"
  tag: "Documentation"
  text: "Configure an API Test"
---

## Overview

Private locations allow you to monitor internal-facing applications or any private URLs that aren’t accessible from the public internet. They can also be used to create a new custom Synthetic Location.

https://cl.ly/2737639336a2
https://cl.ly/2737639336a2/download/Screen%252520Recording%2525202019-08-02%252520at%25252003.09%252520PM.gif

## Setup

The private location worker is shipped as a Docker container, the image can be found here on Dockerhub.

Your private location worker will periodically (by default, every second) pull your test configurations from Datadog’s servers via HTTPS, execute the test depending on the frequency that was defined during the configuration of the test, and return the associated result to Datadog’s servers.
The behavior of private locations is completely identical to the one of managed locations.

> ### Via docker run

Reviewer: reconfigure this section into steps

You first need to create the configuration file associated with your private location on your worker. This file can be found when clicking on Add Private Location on Synthetics / Settings / Private Locations.

https://cl.ly/9be225b3c671
https://cl.ly/9be225b3c671/download/%255B45641e47632b17672af2c9f7493d72f2%255D_Image%2525202019-08-02%252520at%2525207.21.39%252520PM.png

The configuration file contains secrets for private location authentication, test configuration decryption, and test result encryption.
Those secrets are not stored by Datadog, so you should make sure to store them in a safe place before leaving the view. You will need these secrets again, would you decide to add more workers, or to install workers on another host.

> Launch your worker as a standalone container using a Docker run command and the previously created configuration file.

Reviewer: this should be a step with the cmds in copy pasteable code

> Requirements:

Reviewer: ```suggestion
#### Docker requirements
```

Reviewer: should this section be at the beginning of the docker configuration section?

> Docker should be installed on your instance to be able to run the worker.

Reviewer: ```suggestion
- Install Docker on your instance.
```

> The private location worker should be able to access https://api.datadoghq.com/api/ (US instances) or https://api.datadoghq.eu/api/ (EU instances) in order to be able to pull test configurations and push test results.

Reviewer: ```suggestion
- To pull test configurations and push test results, give the private location worker access to one of the Datadog API endpoints: `https://api.datadoghq.com/api/` (for the US instance) or `https://api.datadoghq.eu/api/` (for the EU instance).
```

Reviewer: how do you actually do this?

## Security

> The private location workers only pull data from Datadog servers. Datadog does not push any sort of data to the workers.

Reviewer: ```suggestion
The private location workers only pull data from Datadog servers. Datadog does not push data to the workers.
```

> The secret access key, which is the one used to authenticate your private location worker to Datadog servers, uses an in-house protocol based on AWS Signature Version 4 protocol.

Reviewer: ```suggestion
The secret access key, used to authenticate your private location worker to the Datadog servers, uses an in-house protocol based on AWS Signature Version 4 protocol.
```

> The test configurations are encrypted asymmetrically. The private key will be used to decrypt the test configurations pulled by the workers from Datadog servers. The public key will be used to encrypt the test results that are being sent from the workers to Datadog servers.

Reviewer: ```suggestion
The test configurations are encrypted asymmetrically. The private key is used to decrypt the test configurations pulled by the workers from Datadog servers. The public key is used to encrypt the test results that are sent from the workers to Datadog's servers.
```

## How to scale private locations

> To scale-out private locations, you would need to add additional workers.

Reviewer: ```suggestion
To scale private locations, add or remove workers.
```

Reviewer: I feel like this doesn't need it's own section, just a note in the Docker section

## Liveness of the probe

> You can know more about the liveness of your private locations by looking at the pills displayed:

Reviewer: ```suggestion
For more information about the liveness of your private locations, see the pills displayed:
```

on your private locations list, in the Settings section,
on the form when creating a test, below the Private locations section.

> 

Reviewer: separate these out into a list

> You’ll get a green pill if the private location polled your endpoint less than 5 seconds before loading the settings or create test pages.

Reviewer: is this likely to happen often? my lack of docker expertise is showing here, but is someone setting this up likely to not know what this means?

Author: adding the screenshots helped here

---
title: Run Synthetics tests from Private Locations
kind: documentation
description: Run Synthetics API and browser tests from Private Locations
beta: true
further_reading:
- link: "synthetics/"
  tag: "Documentation"
  text: "Manage your checks"
- link: "synthetics/browser_tests"
  tag: "Documentation"
  text: "Configure a Browser Test"
- link: "synthetics/api_tests"
  tag: "Documentation"
  text: "Configure an API Test"
---

<div class="alert alert-warning">

> This feature is in beta. <a href="https://app.datadoghq.com/privatelocations/2019signup">Sign up for the beta</a> to enable Synthetics private locations for your account.

Reviewer 2: Link to form is different if EU instances - should we specify both?

Author: well spotted.

    {{< img src="synthetics/private_locations/private_location.gif" alt="Private locations" responsive="true" >}}

    **Note**: The configuration file contains secrets for private location authentication, test configuration decryption, and test result encryption. Datadog does not store the secrets, so store them locally before leaving the Private Locations screen.
    **You need to be able to reference these secrets again if you decide to add more workers, or to install workers on another host.**

3. Launch your worker as a standalone container using the Docker run command provided and the previously created configuration file:

    ```
    docker run --name synthetics-private-location-worker-<LOCATION_ID> \
               --rm -v $(pwd)/synthetics-private-location-worker-<LOCATION_ID>.json:/etc/synthetics-private-location-worker-<LOCATION_ID>.json datadog/synthetics-private-location-worker node /datadog-synthetics-agent/dist/worker.js \
               --config=/etc/synthetics-private-location-worker-<LOCATION_ID>.json
    ```

    **Note**: To scale a private location, add or remove workers on your host.

> 4. To pull test configurations and push test results, the private location worker need access to one of the Datadog API endpoints:

Reviewer 2: needS

</div>

## Overview

Private locations allow you to monitor internal-facing applications or any private URLs that aren’t accessible from the public internet. They can also be used to create a new custom Synthetics location.

## Setup

The private location worker is shipped as a Docker container. By default, every second, your private location worker pulls your test configurations from Datadog’s servers using HTTPS, executes the test depending on the frequency defined in the configuration of the test, and returns the test results to Datadog’s servers.

Once you created a private location, configuring a Synthetics API or Browser test from a private location is completely identical to the one of Datadog managed locations.

### Create a new private location

1. Go in *Synthetics* -> *Settings* -> *Private Locations* and create a new private location:

    {{< img src="synthetics/private_locations/create_private_location.png" alt="create a private locations" responsive="true" style="width:70%;">}}

> 2. Get the configuration file associated with your private location on your worker.

Reviewer: ```suggestion
2. Fill out the Location Details and click **Save and Generate** to generate the configuration file associated with your private location on your worker.
```

>     {{< img src="synthetics/private_locations/private_location.gif" alt="Private locations" responsive="true" >}}

Reviewer: worried that this gif shows a secret access key - any way to blur that out?

Author: the access key used is now deleted.. so i would say it's fine imho.. ping @MargotLepizzera

Reviewer 2: Yes, it's fine because the PL was deleted and is not used.
